larky/src/main/resources/stdlib/urllib/request.star

```diff
@@ -1,268 +1,393 @@
+load("@stdlib//types", types="types")
+load("@stdlib//urllib/parse", parse="parse")
+load("@vendor//option/result", Result="Result", Error="Error")
+
+
+__all__ = [
+    # Classes
+    "Request",
+]
+
+
+urlparse = parse.urlparse
+unwrap = parse.unwrap
+_splittag = parse._splittag
+
+
+# copied from cookielib.py
+_cut_port_re = re.compile(r":\d+$")
+
+
+def request_host(request):
+    """Return request-host, as defined by RFC 2965.
+
+    Variation from RFC: returned value is lowercased, for convenient
+    comparison.
+
+    """
+    url = request.full_url
+    host = urlparse(url)[1]
+    if host == "":
+        host = request.get_header("Host", "")
+
+    # remove port, if present
+    host = _cut_port_re.sub("", host, 1)
+    return host.lower()
+
+
+def Request(
+    url=None,
+    data=None,
+    headers=None,
+    origin_req_host=None,
+    unverifiable=False,
+    method=None,
+    **kwargs,
+):
+    if url == None:
+        fail("URL cannot be None")
+    if not headers:
+        headers = {}
+    self = larky.mutablestruct(__name__="Request", __class__=Request)
+
+    # data property
+    def _get_data():
+        return self._data
+
+    def _set_data(data):
+        if data != self._data:
+            self._data = data
+            # issue 16464
+            # if we change data we need to remove content-length header
+            # (cause it's most probably calculated for previous value)
+            if self.has_header("Content-length"):
+                self.remove_header("Content-length")
+    self.data = larky.property(_get_data, _set_data)
+
+    # LarkySanitizedHttpMessage backwards compatibility
+    def body(charset="utf-8"):
+        _body = _get_data()
+        if types.is_bytelike(_body):
+            return _body.decode(charset)
+        return _body
+    self.body = body
+
+    # LarkySanitizedHttpMessage backwards compatibility
+    def set_body(body, charset="utf-8"):
+        if not types.is_bytelike(body):
+            self.data = body.decode(encoding=charset)
+        else:
+            self.data = body
+        return self
+    self.set_body = set_body
+
+    def raw_body():
+        _body = _get_data()
+        if not types.is_bytelike(_body):
+            return codecs.encode(_body, encoding="utf-8")
+        return _body
+    self.raw_body = raw_body
+
+    # full_url property
+    def _get_full_url():
+        if self.fragment:
+            return "{}#{}".format(self._full_url, self.fragment)
+        return self._full_url
+    def _set_full_url(url):
+        # unwrap('<URL:type://host/path>') --> 'type://host/path'
+        self._full_url = unwrap(url)
+        self._full_url, self.fragment = _splittag(self._full_url)
+        self._parse()
+    self.full_url = larky.property(_get_full_url, _set_full_url)
+    # non standard extension (for serialization..)
+    self.url = larky.property(_get_full_url)
+    # LarkySanitizedHttpMessage backwards compatibility
+    self.uri = _get_full_url
```

Comment:
can we make this a property with a getter and a setter? 
you can keep the `set_uri()` method below for backwards compatibility